pygmt/datatypes/header.py

```diff
@@ -0,0 +1,228 @@
+        # Referencing system string in WKT format
+        ("ProjRefWKT", ctp.c_char_p),
+        # Referencing system EPSG code
+        ("ProjRefEPSG", ctp.c_int),
+        # Lower-level information for GMT use only
+        ("hidden", ctp.c_void_p),
+    ]
+
+
+def _parse_nameunits(nameunits: str) -> tuple[str, str | None]:
+    """
+    Get the long_name and units attributes from x_units/y_units/z_units in the grid
+    header.
+
+    In the GMT grid header, the x_units/y_units/z_units are strings in the form of
+    ``long_name [units]``, in which both ``long_name`` and ``units`` are standard
+    netCDF attributes defined by CF conventions. The ``[units]`` part is optional.
+
+    This function parses the x_units/y_units/z_units strings and gets the ``long_name``
+    and ``units`` attributes.
+
+    Parameters
+    ----------
+    nameunits
+        The x_units/y_units/z_units strings in the grid header.
+
+    Returns
+    -------
+    (long_name, units)
+        Tuple of netCDF attributes ``long_name`` and ``units``. ``units`` may be
+        ``None``.
+
+    Examples
+    --------
+    >>> _parse_nameunits("longitude [degrees_east]")
+    ('longitude', 'degrees_east')
+    >>> _parse_nameunits("latitude [degrees_north]")
+    ('latitude', 'degrees_north')
+    >>> _parse_nameunits("x")
+    ('x', None)
+    >>> _parse_nameunits("y")
+    ('y', None)
+    >>>
+    """
+    parts = nameunits.split("[")
+    long_name = parts[0].strip()
+    units = parts[1].strip("]").strip() if len(parts) > 1 else None
+    return long_name, units
+
+
+def _parse_header(header: _GMT_GRID_HEADER) -> tuple[tuple, dict, int, int]:
+    """
+    Get dimension names, attributes, grid registration and type from the grid header.
+
+    For a 2-D grid, the dimension names are set to "x", "y", and "z" by default. The
+    attributes for each dimension are parsed from the grid header following GMT source
+    codes. See the GMT functions "gmtnc_put_units", "gmtnc_get_units" and
+    "gmtnc_grd_info" for reference.
+
+    The last dimension is special and is the data variable name, and the attributes
+    for this dimension are global attributes for the grid.
+
+    The grid is assumed to be Cartesian by default. If the x and y units are
+    "degrees_east" and "degrees_north", respectively, then the grid is assumed to be
+    geographic.
+
+    Parameters
+    ----------
+    header
+        The grid header structure.
+
+    Returns
+    -------
+    dims : tuple
+        The dimension names with the last dimension for the data variable.
+    attrs : dict
+        The attributes for each dimension.
+    registration : int
+        The grid registration. 0 for gridline and 1 for pixel.
+    gtype : int
+        The grid type. 0 for Cartesian grid and 1 for geographic grid.
+    """
+    # Default dimension names. The last dimension is for the data variable.
+    dims: tuple = ("x", "y", "z")
+    nameunits = (header.x_units, header.y_units, header.z_units)
```

Comment:
Reorder to yxz as mentioned at https://github.com/GenericMappingTools/pygmt/pull/3134#discussion_r1535769435
```suggestion
    dims: tuple = ("y", "x", "z")
    nameunits = (header.y_units, header.x_units, header.z_units)
```